### PR TITLE
Use inline Vector{128|256}.Create for constants

### LINKED ISSUE #44115: Revise how constant SIMD vectors are defined in BCL
_From https://github.com/dotnet/runtime/pull/44111#discussion_r515546381_

There are 3 patterns we currently use across the BCL for const vectors:
```csharp
    //
    // Case 1: Plain Vector.Create
    public Vector128<byte> Case1(Vector128<byte> vec)
    {
        Vector128<byte> mask = Vector128.Create(
            0xFF, 0xFF, 0, 0xFF, 0xFF, 0xFF, 1, 0xFF,
            0xFF, 0xFF, 2, 0xFF, 0xFF, 0xFF, 3, 0xFF);

        return Ssse3.Shuffle(vec, mask);
    }


    //
    // Case 1.1: Plain Vector.Create as argument of some SIMD instruction directly
    // Should be the same codegen as for Case1 ^ (spoiler: it's not. Forward Substitution? see #4655)
    public Vector128<byte> Case1_1(Vector128<byte> vec)
    {
        return Ssse3.Shuffle(vec,
                    Vector128.Create( // used without "mask" local as in Case1 ^
                        0xFF, 0xFF, 0, 0xFF, 0xFF, 0xFF, 1, 0xFF,
                        0xFF, 0xFF, 2, 0xFF, 0xFF, 0xFF, 3, 0xFF));
    }


    //
    // Case 2: static readonly Vector
    private static readonly Vector128<byte> s_mask = Vector128.Create(
        0xFF, 0xFF, 0, 0xFF, 0xFF, 0xFF, 1, 0xFF,
        0xFF, 0xFF, 2, 0xFF, 0xFF, 0xFF, 3, 0xFF);
    public Vector128<byte> Case2(Vector128<byte> vec)
    {
        // we also often save it to a local first (e.g. before loops)
        return Ssse3.Shuffle(vec, s_mask);
    }


    //
    // Case 3: Roslyn's hack
    private static ReadOnlySpan<byte> Mask => new byte[] {
        0xFF, 0xFF, 0, 0xFF, 0xFF, 0xFF, 1, 0xFF,
        0xFF, 0xFF, 2, 0xFF, 0xFF, 0xFF, 3, 0xFF };
    public Vector128<byte> Case3(Vector128<byte> vec)
    {
        return Ssse3.Shuffle(vec, Unsafe.ReadUnaligned<Vector128<byte>>(
            ref MemoryMarshal.GetReference(Mask)));
    }
```
Here is the current codegen for these cases:
```asm
; Method Case1
G_M46269_IG01:
       vzeroupper 
G_M46269_IG02:
       vmovupd  xmm0, xmmword ptr [reloc @RWD00]        ; loaded from the data section, OK
       vmovupd  xmm1, xmmword ptr [r8]
       vpshufb  xmm0, xmm1, xmm0
       vmovupd  xmmword ptr [rdx], xmm0
       mov      rax, rdx
G_M46269_IG03:
       ret      
RWD00  	dq	FF01FFFFFF00FFFFh, FF03FFFFFF02FFFFh    ; <-- it's here
; Total bytes of code: 29



; Method Case1_1
G_M7091_IG01:
       vzeroupper 
G_M7091_IG02:
       vmovupd  xmm0, xmmword ptr [r8]
       vpshufb  xmm0, xmm0, xmmword ptr [reloc @RWD00]  ; loaded as part of vshufb without 
                                                        ; additional registers from the data secion - PERFECT!!
       vmovupd  xmmword ptr [rdx], xmm0
       mov      rax, rdx
G_M7091_IG03:
       ret      
RWD00  	dq	FF01FFFFFF00FFFFh, FF03FFFFFF02FFFFh    ; <-- it's here
; Total bytes of code: 25



; Method Case2
G_M31870_IG01:
       push     rsi
       sub      rsp, 48
       vzeroupper 
       mov      rsi, rdx
G_M31870_IG02:
       vmovupd  xmm0, xmmword ptr [r8]
       vmovupd  xmmword ptr [rsp+20H], xmm0
       mov      rcx, 0xD1FFAB1E
       mov      edx, 2
       call     CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE  ; static initialization (in some cases can be eliminated by jit but still...)
       mov      rax, 0xD1FFAB1E                          ; <- additional mov
       mov      rax, gword ptr [rax]
       vmovupd  xmm0, xmmword ptr [rsp+20H]
       vpshufb  xmm0, xmm0, xmmword ptr [rax+8]
       vmovupd  xmmword ptr [rsi], xmm0
       mov      rax, rsi
G_M31870_IG03:
       add      rsp, 48
       pop      rsi
       ret      
; Total bytes of code: 80



; Method Case3
G_M23615_IG01:
       vzeroupper 
G_M23615_IG02:
       mov      rax, 0xD1FFAB1E                          ; <- additional mov
                                                         ; kind of makes sense if the same mask is used from different methods
                                                         ; but the C# code looks a bit ugly
       vmovupd  xmm0, xmmword ptr [rax]
       vmovupd  xmm1, xmmword ptr [r8]
       vpshufb  xmm0, xmm1, xmm0
       vmovupd  xmmword ptr [rdx], xmm0
       mov      rax, rdx
G_M23615_IG03:
       ret      
; Total bytes of code: 35
```
The first case used to be avoided due to some codegen issues, but looks like those were resolved (e.g. JIT now saves such vectors into the data section, does Value Numbering for SIMDs including constant vectors, does CSE, etc - https://github.com/dotnet/runtime/pull/31834?) so we now have a lot of `static readonly` fields and we can revise them and convert into `Case1(1.1)`-style where possible (maybe even if we need to duplicate them in different methods), e.g.:

Places to revise:

1. [Base64Encoder.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Memory/src/System/Buffers/Text/Base64Encoder.cs#L525-L582)
2. [Base64Decoder.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Memory/src/System/Buffers/Text/Base64Decoder.cs#L616-L716)
3. [Sse2Helper.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Text.Encodings.Web/src/System/Text/Encodings/Web/Sse2Helper.cs#L103-L119)
4. [Ssse3Helper.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Text.Encodings.Web/src/System/Text/Encodings/Web/Ssse3Helper.cs#L103-L106)
5. [BitArray.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Collections/src/System/Collections/BitArray.cs#L116)
6. Maybe there are more
7. Scan other repos: aspnetcore, ML.NET, etc

Known limitations for Case1:
- See Case1.1 comment
- JIT doesn't hoist Vector.Create from loops' bodies yet (would be nice to have)

/cc @stephentoub @GrabYourPitchforks @benaadams @tannergooding

### PR DESCRIPTION
**Pull request by gfoidl**
Fixes https://github.com/dotnet/runtime/issues/44115

Used `Vector(128|256)<\w+> \w+ =` as regex-search, seems that these three places are the onley one left.
The SSE-helper, that are mentioned in the issue, got re-written by @GrabYourPitchforks some time ago, and there patterns for the new codegen is already used.

Codegen was validated (manually) to see that it's an improvement.

It's not clear to me whether the touched codes are built for .NET Core 3.1 or not. So I added conditional compilation, as for .NET Core 3.1 the codegen is quite worse when Vector{128|256}.Create is used (that's why we initially had the different patterns before the JIT started to emit constants and read them from the data-segment).

/cc: @tannergooding

### CODE REVIEW

File: src/libraries/System.Collections/src/System/Collections/BitArray.cs [hunk 1]
```diff
@@ -120,9 +120,11 @@ public BitArray(byte[] bytes)
             _version = 0;
         }
 
+#if !NET6_0_OR_GREATER
```
Comment (stephentoub) [inline src/libraries/System.Collections/src/System/Collections/BitArray.cs:123]: Why are these ifdefs needed... isn't this file only compiled for netcoreappcurrent?

File: src/libraries/System.Memory/src/System/Buffers/Text/Base64Decoder.cs [hunk 2]
```diff
@@ -362,14 +362,64 @@ private static unsafe void Avx2Decode(ref byte* srcBytes, ref byte* destBytes, b
             // See SSSE3-version below for an explanation of how the code works.
 
             // The JIT won't hoist these "constants", so help it
+#if NET6_0_OR_GREATER
```
Comment (stephentoub) [inline src/libraries/System.Memory/src/System/Buffers/Text/Base64Decoder.cs:365]: Same question

File: src/libraries/System.Memory/src/System/Buffers/Text/Base64Encoder.cs [hunk 3]
```diff
@@ -241,14 +241,36 @@ private static unsafe void Avx2Encode(ref byte* srcBytes, ref byte* destBytes, b
             // l k j i h g f e d c b a 0 0 0 0
 
             // The JIT won't hoist these "constants", so help it
+#if NET6_0_OR_GREATER
```
Comment (stephentoub) [inline src/libraries/System.Memory/src/System/Buffers/Text/Base64Encoder.cs:244]: Ditto.  I don't think any of these ifdefs are needed.

File: src/libraries/System.Collections/src/System/Collections/BitArray.cs (hunk 1 above)
Comment (gfoidl) [reply src/libraries/System.Collections/src/System/Collections/BitArray.cs:123]: > isn't this file only compiled for netcoreappcurrent?

This is the point I don't know (alluded in the description).
But if so, I'll just remove them.

My point is just to make sure that .NET Core 3.1 doesn't regress, as it's a LTS through Dec, 2022.

Comment (stephentoub) [reply src/libraries/System.Collections/src/System/Collections/BitArray.cs:123]: They're not needed.  This file is only included in System.Collections.csproj, which only builds for netcoreappcurrent:
https://github.com/dotnet/runtime/blob/bde0322a9e11fba40a7a65c6d38200d6514efc72/src/libraries/System.Collections/src/System.Collections.csproj#L5

Base64Decoder/Encoder.cs are only included in System.Memory, which only builds for netcoreappcurrent:
https://github.com/dotnet/runtime/blob/bde0322a9e11fba40a7a65c6d38200d6514efc72/src/libraries/System.Memory/src/System.Memory.csproj#L6

Comment (gfoidl) [reply src/libraries/System.Collections/src/System/Collections/BitArray.cs:123]: Thanks! 👍🏻

**Review by tannergooding — APPROVED** [review]

**Pull request merged**